### PR TITLE
[997] Authorisation with pundit

### PR DESCRIPTION
**Pull request by benilovj**
### Context
To prevent unauthorised publishers inadvertently gaining access to other publishers' data.

### Changes proposed in this pull request
Introduce [Pundit](https://github.com/varvet/pundit), a commonly used authorisation framework and apply it to all `v1` and `v2` endpoints.

### CODE REVIEW

File: spec/requests/api/v2/providers/courses_spec.rb [hunk 1]
```diff
@@ -18,20 +19,33 @@
...
+    }
     subject { response }
 
     before do
       get "/api/v2/providers/#{provider.provider_code}/courses",
           headers: { 'HTTP_AUTHORIZATION' => credentials }
     end
 
-    context 'when unauthorized' do
+    context 'when unauthenticated' do
```
Comment (defong) [inline spec/requests/api/v2/providers/courses_spec.rb]: `unauthenticated` vs `unauthorized` you sure as http code returns `unauthorized`

File: Gemfile [hunk 2]
```diff
@@ -9,6 +9,9 @@ gem 'rails', '~> 5.2.2'
 # Use PostgreSQL as the database for Active Record
 gem 'pg'
 
+# Authorisation
+gem "pundit"
```
Comment (dankmitchell) [inline Gemfile]: Single quotes / double 🤷‍♂️

File: app/controllers/api/v2/application_controller.rb [hunk 3]
```diff
@@ -1,6 +1,8 @@
 module API
   module V2
     class ApplicationController < ::ApplicationController
+      attr_reader :current_user
```
Comment (dankmitchell) [inline app/controllers/api/v2/application_controller.rb]: I wonder if we could use the new current attributes from rails 5.2 here?

File: spec/requests/api/v2/providers/courses_spec.rb (hunk 1 above)
Comment (benilovj) [reply spec/requests/api/v2/providers/courses_spec.rb]: I **think** this is right – there isn't a separate `unauthorized` HTTP status code to my knowledge.

As I understand it:

* an unauthenticated user isn't someone who we have on the system
* an unautorised user is someone who we have on the system, but shouldn't have access to the thing they're trying to access

File: Gemfile (hunk 2 above)
Comment (benilovj) [reply Gemfile]: Good spot, fixed.

File: app/controllers/api/v2/application_controller.rb (hunk 3 above)
Comment (benilovj) [reply app/controllers/api/v2/application_controller.rb]: I've not come across this yet – anything you can point me towards?

Comment (dankmitchell) [reply app/controllers/api/v2/application_controller.rb]: Create a class or concern like so -

```
class Current < ActiveSupport::CurrentAttributes
  attribute :user
end
```

then you can assign `Current.user = @user` anywhere which gives you access to `current_user` throughout the app.

Haven't used it myself, just read about it.

Comment (dankmitchell) [reply app/controllers/api/v2/application_controller.rb]: Don't let this stop you merging...

File: spec/requests/api/v2/providers/courses_spec.rb (hunk 1 above)
Comment (misaka) [reply spec/requests/api/v2/providers/courses_spec.rb]: Ya, I think you're right. And as you point out, @benilovj, "Authorisation" seems to be the catch-all term, but if we can disambiguate, that seems to be the correct definition.

File: app/policies/provider_policy.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class ProviderPolicy
```
Comment (misaka) [inline app/policies/provider_policy.rb]: The other thing we should implement here is the [policy scope](https://github.com/varvet/pundit#scopes). Were you saving this for a later PR?

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 
 private
 
   def render_service_unavailable
     render json: { code: 503, status: 'Service Unavailable' }.to_json, status: 503
   end
+
+  def authorization_error
+    render json: { code: 401, status: 'Unauthorised user' }.to_json, status: 401
+  end
```
[the remark this answers is not in the document]
Comment (defong) [reply app/controllers/application_controller.rb]: see https://github.com/DFE-Digital/manage-courses-backend/pull/124 as example

**Comment by benilovj** [comment]
@misaka I've had to tweak b147811da07d193a8025ce8cc33eb3ac70ddcd9e to get a couple of request specs to pass, they broke as a result of the change to exception handling. It's running green now...

**Pull request merged**